File: src/ai_sdr_agent/services/tool_config.py
from __future__ import annotations

import re
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

HttpMethod = Literal["GET", "POST", "PUT", "PATCH", "DELETE"]
PreToolSpeech = Literal["auto", "force", "disabled"]
ExecutionMode = Literal["default", "blocking"]
ToolCallSound = Literal["none", "click", "custom_url"]
AuthType = Literal["none", "bearer", "basic", "api_key_header", "connection"]


class ToolHeader(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    value: str = ""


class ToolParameterDef(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    description: str = ""
    type: Literal["string", "number", "integer", "boolean"] = "string"
    required: bool = False


class ToolAuthConfig(BaseModel):
    type: AuthType = "none"
    bearer_token: str | None = None
    basic_username: str | None = None
    basic_password: str | None = None
    api_key_header_name: str | None = None
    api_key_value: str | None = None
    connection_id: str | None = None


class HttpToolConfigV1(BaseModel):
    schema_version: Literal[1] = 1
    method: HttpMethod = "POST"
    url: str = ""
    response_timeout_seconds: int = Field(default=20, ge=1, le=120)
    disable_interruptions: bool = False
    pre_tool_speech: PreToolSpeech = "auto"
    pre_tool_speech_text: str | None = None
    execution_mode: ExecutionMode = "default"
    tool_call_sound: ToolCallSound = "none"
    tool_call_sound_url: str | None = None
    auth: ToolAuthConfig = Field(default_factory=ToolAuthConfig)
    headers: list[ToolHeader] = Field(default_factory=list)
    path_parameters: list[ToolParameterDef] = Field(default_factory=list)
    query_parameters: list[ToolParameterDef] = Field(default_factory=list)
    parameters: dict[str, Any] | None = None

    @field_validator("url")
    @classmethod
    def url_not_empty_when_saved(cls, v: str) -> str:
        return v.strip()


_ENV_PATTERN = re.compile(r"\{\{([A-Za-z_][A-Za-z0-9_]*)\}\}")


def extract_path_param_names(url: str) -> list[str]:
    return list(dict.fromkeys(re.findall(r"\{([a-zA-Z_][a-zA-Z0-9_]*)\}", url)))


def parse_tool_config(raw: dict[str, Any] | None) -> HttpToolConfigV1:
    if not raw:
        return HttpToolConfigV1()
    data = dict(raw)
    if data.get("schema_version") != 1:
        migrated = _migrate_legacy_config(data)
        return HttpToolConfigV1.model_validate(migrated)
    return HttpToolConfigV1.model_validate(data)


def _migrate_legacy_config(data: dict[str, Any]) -> dict[str, Any]:
    url = data.get("url") or data.get("endpoint_url") or ""
    method = (data.get("method") or "POST").upper()
    if method not in ("GET", "POST", "PUT", "PATCH", "DELETE"):
        method = "POST"
    return {
        "schema_version": 1,
        "method": method,
        "url": url,
        "response_timeout_seconds": 20,
    }


def llm_function_name(tool_id: str, display_name: str) -> str:
    slug = re.sub(r"[^a-zA-Z0-9_]+", "_", display_name.strip().lower()).strip("_")
    if not slug:
        slug = "tool"
    short_id = str(tool_id).replace("-", "")[:8]
    name = f"tool_{short_id}_{slug}"
    if len(name) > 64:
        name = name[:64].rstrip("_")
    return name

File: vocal-frontend/src/lib/toolConfig.ts
export type HttpMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE";
export type PreToolSpeech = "auto" | "force" | "disabled";
export type ExecutionMode = "default" | "blocking";
export type ToolCallSound = "none" | "click" | "custom_url";
export type AuthType = "none" | "bearer" | "basic" | "api_key_header" | "connection";

export interface ToolHeader {
  name: string;
  value: string;
}

export interface ToolParameterDef {
  name: string;
  description: string;
  type: "string" | "number" | "integer" | "boolean";
  required: boolean;
}

export interface ToolAuthConfig {
  type: AuthType;
  bearer_token?: string | null;
  basic_username?: string | null;
  basic_password?: string | null;
  api_key_header_name?: string | null;
  api_key_value?: string | null;
  connection_id?: string | null;
}

export interface HttpToolConfigV1 {
  schema_version: 1;
  method: HttpMethod;
  url: string;
  response_timeout_seconds: number;
  disable_interruptions: boolean;
  pre_tool_speech: PreToolSpeech;
  pre_tool_speech_text?: string | null;
  execution_mode: ExecutionMode;
  tool_call_sound: ToolCallSound;
  tool_call_sound_url?: string | null;
  auth: ToolAuthConfig;
  headers: ToolHeader[];
  path_parameters: ToolParameterDef[];
  query_parameters: ToolParameterDef[];
  parameters?: Record<string, unknown> | null;
}

export function defaultHttpToolConfig(): HttpToolConfigV1 {
  return {
    schema_version: 1,
    method: "POST",
    url: "",
    response_timeout_seconds: 20,
    disable_interruptions: false,
    pre_tool_speech: "auto",
    pre_tool_speech_text: null,
    execution_mode: "default",
    tool_call_sound: "none",
    tool_call_sound_url: null,
    auth: { type: "none" },
    headers: [],
    path_parameters: [],
    query_parameters: [],
    parameters: {
      type: "object",
      properties: {},
      required: [],
      additionalProperties: false,
    },
  };
}

export function parseToolConfig(raw: Record<string, unknown> | null | undefined): HttpToolConfigV1 {
  if (!raw || Object.keys(raw).length === 0) {
    return defaultHttpToolConfig();
  }
  if (raw.schema_version === 1) {
    return { ...defaultHttpToolConfig(), ...(raw as unknown as HttpToolConfigV1), schema_version: 1 };
  }
  const url = typeof raw.url === "string" ? raw.url : typeof raw.endpoint_url === "string" ? raw.endpoint_url : "";
  const method = (typeof raw.method === "string" ? raw.method : "POST").toUpperCase() as HttpMethod;
  return {
    ...defaultHttpToolConfig(),
    method: ["GET", "POST", "PUT", "PATCH", "DELETE"].includes(method) ? method : "POST",
    url,
  };
}

export function configToJson(config: HttpToolConfigV1): Record<string, unknown> {
  return { ...config };
}

export function extractPathParamNames(url: string): string[] {
  const matches = url.matchAll(/\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g);
  const names: string[] = [];
  for (const m of matches) {
    if (!names.includes(m[1])) names.push(m[1]);
  }
  return names;
}

export function syncPathParameters(url: string, existing: ToolParameterDef[]): ToolParameterDef[] {
  const names = extractPathParamNames(url);
  const byName = new Map(existing.map((p) => [p.name, p]));
  return names.map(
    (name) =>
      byName.get(name) ?? {
        name,
        description: "",
        type: "string",
        required: true,
      },
  );
}

export function validateParametersJson(text: string): string | null {
  try {
    const parsed = JSON.parse(text) as Record<string, unknown>;
    if (parsed.type !== "object") {
      return "JSON Schema must have type: object";
    }
    return null;
  } catch {
    return "Invalid JSON";
  }
}
